### PR TITLE
feat: Distill the `EvaluationProgram`

### PR DESCRIPTION
**Pull request by winston-h-zhang**
This PR adds the distilled program and distilled memory to loam. Here are the changes:
1. Implemented `cons/car/cdr/atom/quote`, as well as tests.
2. Implemented the distilled program.
3. Implemented the memory distillation process.
4. Added tests for inconsistent memory.
5. Added debugging and printing infrastructure for `ascent`, and for memory distillation.

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -31,7 +31,7 @@ p3-util = { git = "https://github.com/argumentcomputer/Plonky3.git", branch = "s
 sphinx-core = { git = "https://github.com/argumentcomputer/sphinx.git", branch = "dev"}
 sphinx-derive = { git = "https://github.com/argumentcomputer/sphinx.git", branch = "dev" }
 anyhow = "1.0.72"
-ascent = "0.6.0"
+ascent = { git = "https://github.com/winston-h-zhang/ascent.git", branch = "master" }
```
[the remark this answers is not in the document]
Comment (winston-h-zhang) [reply Cargo.toml]: Done, it's moved to https://github.com/argumentcomputer/ascent.

File: src/loam/mod.rs [hunk 2]
```diff
@@ -48,11 +62,47 @@ impl Ptr {
         Self(Tag::Nil.elt(), LE::from_canonical_u32(0))
     }
 
-    fn t() -> Self {
-        let addr = lurk_sym_index("t").unwrap();
+    /// make this const
+    fn buitin(op: &str) -> Self {
```
Comment (porcuquine) [inline src/loam/mod.rs:66]: builtin

File: src/loam/mod.rs [hunk 3]
```diff
@@ -83,6 +133,62 @@ impl Ptr {
...
+        }
+    }
+}
+
+/// Possible ways two pointer can be "equal" to each other
+#[derive(Clone, Copy, Debug, Ord, PartialOrd, PartialEq, Eq, Hash)]
+pub enum PtrEq {
+    Equal = 0,
+    NotEqual,
+    DontKnow,
```
Comment (porcuquine) [inline src/loam/mod.rs:174]: How about `Unknown`?

File: src/loam/allocation.rs [hunk 4]
```diff
@@ -346,174 +516,164 @@ ascent! {
...
+
+        // Start the second phase
+        let mut prog = DistilledAllocationProgram::default();
+
+        // Move the hashes (stored in the allocator) and input over
+        prog.allocator = original_program.allocator.clone();
+        prog.input_expr = original_program.input_expr.clone();
+
+        // transfer over the memory (assume it's been distilled)
+        let raw_memory = original_program.export_memory();
```
Comment (porcuquine) [inline src/loam/allocation.rs:624]: Eventually, it would be good to introduce a new type for the distilled memory so we don't have to assume.

File: src/loam/allocation.rs [hunk 5]
```diff
@@ -346,174 +516,164 @@ ascent! {
...
+        let mut store = Store::default();
+        let options = DistillationOptions::new().with_summary(0.9);
+        let memory = raw_memory.distill_with_store(&mut store, &options);
```
Comment (porcuquine) [inline src/loam/allocation.rs:627]: Is the comment above wrong? If this was already distilled, why are we distilling here? If wrong, let's remove the comment on line 624.

File: src/loam/allocation.rs [hunk 6]
```diff
@@ -346,174 +516,164 @@ ascent! {
...
+            let bad_zptr = zstore.read(bad_input).expect("failed to read");
+            (
+                store.zptr_ptr(&bad_zptr).unwrap(),
+                vec![(read_wideptr(&mut zstore, bad_output),)],
+            )
+        } else {
+            let input_zptr = prog.input_expr[0].0.to_zptr();
+            (
+                store.zptr_ptr(&input_zptr).unwrap(),
+                original_program.output_expr,
```
Comment (porcuquine) [inline src/loam/allocation.rs:640]: What is actually being tested here? Is it the possibility that the `unwrap`s might fail? I would expect to see some assertion performed using these results.

Comment (porcuquine) [reply src/loam/allocation.rs:640]: I understand what's going on now, but it's hard to follow. I think a little documentation of how this is meant to be used would help.

File: src/loam/memory.rs [hunk 7]
```diff
@@ -0,0 +1,752 @@
...
+    fn addr(&self) -> LE {
+        self.0 .1
+    }
+}
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+pub enum PPtrType {
+    Tuple2(PPtr, PPtr),
+    Tuple3(PPtr, PPtr, PPtr),
+}
```
Comment (porcuquine) [inline src/loam/memory.rs:111]: This struct is confusingly named. Can you come up with something that doesn't make usage appear to be naming a *type*?

Comment (winston-h-zhang) [reply src/loam/memory.rs:111]: I'll rename to `PPtrKind`. For more context, it's following the convention with `ZPtr`/`ZPtrType`, [here](https://github.com/argumentcomputer/lurk/blob/bfe044b80854b51e128317962a4b2ec52eb555fa/src/lurk/zstore.rs#L228-L238).

File: src/loam/allocation.rs (hunk 4 above)
Comment (winston-h-zhang) [reply src/loam/allocation.rs:624]: Sorry for the confusing comment, it should've been deleted. There are two types of memory: `VirtualMemory` and `Memory`. The first is the output of the 1st phase, the second is the input to the 2nd phase. 

The behavior of the 1st phase is described by the `LoamProgram` trait and you can implement how to `export_memory`. You have to manually implement `import_memory` for the 2nd phase to be able to import memory. I think the trait organization could be better, but for now you shouldn't be able to accidentally export distilled memory, or import undistilled memory.

**Comment by winston-h-zhang** [comment]
I'm going to rename `test_distilled` to `test_second_phase` and clean up the testing code slightly.

**Review by porcuquine — APPROVED** [review]
This looks good to me. I didn't scrutinize the actual evaluation programs closely again, but I did review them over time as the work built up. Since that's all in flux anyway, I think this is good to merge now.

Let's try to keep PRs smaller, more focused, and less long-lived now that we're in the next phase of development. Good work.

**Review by porcuquine — APPROVED** [review]

**Comment by porcuquine** [comment]
🎉
